fedora/client/openidcclient.py

```diff
@@ -0,0 +1,387 @@
+                    token = self._cache[uuid]
+                else:
+                    # This token was invalid and could not be refreshed
+                    self._delete_token(uuid)
+                    continue
+            # We either refreshed it, 
+            self._valid_cache.append(uuid)
+            return token
+
+    def _get_server(self, app):
+        """This function returns a SimpleServer with an available WEB_PORT."""
+        for port in WEB_PORTS:
+            try:
+                server = simple_server.make_server('0.0.0.0', port, app)
+                return server
+            except socket.error:
+                # This port did not work. Switch to next one
+                continue
+
+    def _get_new_token(self, scopes):
+        """This function kicks off some magic.
+
+        We will start a new webserver on one of the WEB_PORTS, and then either
+        show the user a URL, of if possible, kick off their browser.
+        This URL will be the Authorization endpoint of the IdP with a request
+        for our client_id to get a new token with the specified scopes.
+        The webserver will then need to catch the return with either an
+        Authorization Code (that we will exchange for an access token) or the
+        cancellation message.
+
+        This function will store the new token in the local cache, add it to
+        the valid cache, and then return the UUID.
+        If the user cancelled (or we got another error), we will return None.
+        """
+        def token_app(environ, start_response):
+            query = environ['QUERY_STRING']
+            split = query.split('&')
+            kv = dict([v.split('=', 1) for v in split])
+
+            if 'error' in kv:
+                self.debug('Error code returned: %s (%s)'
+                           % (kv['error'], kv.get('error_description')))
+                self.__retrieved_code = False
+            else:
+                self.__retrieved_code = kv['code']
+
+            # Just return a message
+            out = StringIO()
+            out.write('You can close this window and return to the CLI')
+            start_response('200 OK', [('Content-Type', 'text/plain')])
+            return [out.getvalue()]
+
+        self.__retrieved_code = None
+        server = self._get_server(token_app)
+        if not server:
+            raise Exception('We were unable to instantiate a webserver')
+        return_uri = 'http://localhost:%i/' % server.socket.getsockname()[1]
+        query = {}
+        # We always silently inject scope openid so we get an id_token
+        if len(scopes) > 0:
+            query['scope'] = 'openid %s' % ' '.join(scopes)
+        else:
+            query['scope'] = 'openid'
+        query['response_type'] = 'code'
+        query['client_id'] = self.client_id
+        query['redirect_uri'] = return_uri
+        query['state'] = 'ignored'
+        query['response_mode'] = 'query'
+        query = urllib.urlencode(query)
```

Comment:
not gonna work on py3.. it's in urllib.parse there. you can use six.moves to get urllib_parse